food-truck-api/src/main/java/food/truck/api/reviews_and_subscriptions/ReviewService.java

```diff
@@ -1,26 +1,92 @@
 package food.truck.api.reviews_and_subscriptions;
 
+import food.truck.api.truck.Truck;
+import food.truck.api.truck.TruckRepository;
+import food.truck.api.user.User;
+import lombok.extern.log4j.Log4j2;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
+import java.sql.Timestamp;
+import java.time.Instant;
 import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
+@Log4j2
 @Service
 public class ReviewService {
     @Autowired
     private ReviewRepository reviewRepository;
 
-    public List<Review> findReviewsByUserId(Long userId){
+    @Autowired
+    private TruckRepository truckRepository;
+
+    public List<Review> findReviewsByUserId(Long userId) {
         return reviewRepository.findByUserId(userId);
     }
 
-    public List<Review> findReviewByTruckId(Long truckId){
+    public List<Review> findReviewByTruckId(Long truckId) {
         return reviewRepository.findByTruckId(truckId);
     }
 
-    public Review saveReview(Review review){
-        return reviewRepository.save(review);
+    public Review findReviewById(long reviewId) {
+        Optional<Review> r = reviewRepository.findById(reviewId);
+        return r.orElse(null);
+    }
+
+    public Review saveReview(Long userId, Truck t, int score, int costRating, String reviewText) {
+        // Get the current review, if it exists, or create a new one
+        Optional<Review> rOpt = reviewRepository.findByUserIdAndTruckId(userId, t.getId()).stream().findFirst();
+        Review r = rOpt.orElse(new Review());
+
+        // Populate the review
+        r.setUserId(userId);
+        r.setTruck(t);
+        r.setStarRating(score);
+        r.setCostRating(costRating);
+        r.setReviewText(reviewText);
+        r.setDay_time(Timestamp.from(Instant.now()));
+        var result = reviewRepository.save(r);
+        updateAverageRating(result);
+        return result;
+    }
+
+    public boolean deleteReviewByUser(User u, long truckId) {
+        Review r = reviewRepository.findByUserIdAndTruckId(u.getId(), truckId).stream().findFirst().orElse(null);
+        if (r == null)
+            return false; // TODO make return a 404
+        else if (!r.getUserId().equals(u.getId()))
+            return false; // TODO make return a 403
+
+        // Delete review
+        reviewRepository.delete(r);
+        updateAverageRating(r);
+        return true;
+    }
+
+    public Review findReviewForUser(Long userId, Long truckId) {
+        return reviewRepository.findByUserIdAndTruckId(userId, truckId).stream().findFirst().orElse(null);
     }
 
+    /**
+     * I hate this solution, but it didn't work out any other way...
+     */
```

Comment:
Is this because you don't want to look at every single review to update average? I think it would work if you do a weighted average, like newAvg = (oldAvg * numReviews + newReview) / (numReviews + 1). But it would probably be less error-prone to just recalculate it.

Reply:
The reason was because I know that this means we need to keep track of anywhere we want to update reviews and run this code there, which is asking for a bug (where we update reviews and forget to run it). I wanted to set it up as a trigger-ish thing on updating the Reviews table, but Hibernate did not want to cooperate.

Reply:
This sounds like a good way to me, only allowing changes to the reviews through the public methods in ReviewService, and updating the average in those methods